### PR TITLE
Fixes #11402 - Add defaults support for location/organization

### PR DESCRIPTION
**Pull request by alongoldboim**
This is the first experiment with the defaults feature.
For now only default organization/location is working.

### CODE REVIEW

**Comment by alongoldboim** [comment]
Fails because of a change that depends on a change in the hammer_cli that hasn't entered yet.

File: lib/hammer_cli_foreman/defaults.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+module HammerCLIForeman
+
+  class Defaults < HammerCLI::Defaults
+    SUPPORTED_DEFAULTS=[:organization_id, :location_id]
+    def self.register_provider
+      super(self)
+    end
+
+    def self.support?(param)
```
Comment (tstrachota) [inline lib/hammer_cli_foreman/defaults.rb]: :+1: nice pattern!

File: lib/hammer_cli_foreman/defaults.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      SUPPORTED_DEFAULTS.any? {|s| s.to_s == param}
+    end
+
+    def self.get_defaults(param)
+      param = "default_organization" if param == :organization_id
+      param = "default_location" if param == :location_id
+      user = HammerCLIForeman.foreman_resource(:users).action(:index).call(:search =>"login="+HammerCLI::Settings.settings[:foreman][:username])
+      user_id = user["results"][0]["id"] if user["results"]
+      values = HammerCLIForeman.foreman_resource(:users).action(:show).call(:id => user_id)
```
Comment (tstrachota) [inline lib/hammer_cli_foreman/defaults.rb]: Default organization and location are returned already in the index call so there's probably no need to do the second call here.

File: lib/hammer_cli_foreman/defaults.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+require 'hammer_cli'
+module HammerCLIForeman
+
+  class Defaults < HammerCLI::Defaults
```
Comment (tstrachota) [inline lib/hammer_cli_foreman/defaults.rb]: Do we need to inherit from the main defaults manager class? I don't see much connection here.

Comment (alongoldboim) [inline lib/hammer_cli_foreman/defaults.rb]: I implemented it that way before we changed to register providers, now we don't really need it.
ill refactor it and update you.
Thanks.

**Comment by alongoldboim** [comment]
@mbacovsky @tstrachota we need to think what we want to implement here and what defaults we want to support, this plugin will serve as a base template for other plugins we would like to serve defaults from.
For now its workingl for org/loc.

**Comment by alongoldboim** [comment]
Another issue we have is that the server calls wont be able to be answered unless user password are set on the conf file.

File: lib/hammer_cli_foreman/defaults.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    def self.supported_defaults
+      SUPPORTED_DEFAULTS
+    end
+
+    def self.get_defaults(param)
+      param = "default_organization" if param == :organization_id
+      param = "default_location" if param == :location_id
+      user = HammerCLIForeman.foreman_resource(:users).action(:index).call(:search =>"login="+HammerCLI::Settings.settings[:foreman][:username]) if HammerCLI::Settings.settings[:foreman][:username]
```
Comment (mbacovsky) [inline lib/hammer_cli_foreman/defaults.rb]: To address the issue with missing username you can use HammerCLIForeman.credentials that ask interactively if necessary:
```ruby
      user = HammerCLIForeman.foreman_resource(:users).action(:index).call(:search =>"login="+HammerCLIForeman.credentials.username)
```

**Comment by mbacovsky** [comment]
I think for now organization and location are good to start with. These are the only ones requested in the BZ, we can add more later according to demand.

File: lib/hammer_cli_foreman/defaults.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    private
+    def get_user
+      HammerCLIForeman.foreman_resource(:users).action(:index).call(:search =>"login="+HammerCLIForeman.credentials.username)
+    end
+  end
+  HammerCLI.defaults.register_provider(Defaults.new())
+
+end
```
Comment (mbacovsky) [inline lib/hammer_cli_foreman/defaults.rb]: no newline at the end of file

File: test/unit/defaults_test.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
+require File.join(File.dirname(__FILE__), 'test_helper')
+describe HammerCLIForeman::Defaults do
+
+  context "Defaults" do
+
+    defaults_provider = HammerCLIForeman::Defaults.new
+    user = {"results" => ["default_organization" => {"id" => 1}, "default_location" => {"id" => 1}]}
```
Comment (mbacovsky) [inline test/unit/defaults_test.rb]: could you use different id values to avoid possibility overseeing of swapping them by mistake?

**Comment by mbacovsky** [comment]
Besides the little things mentioned above it looks good. 
Could you please rebase?

**Comment by mbacovsky** [comment]
:+1: Good job, @alongoldboim! :beers:

**Comment by tstrachota** [comment]
@alongoldboim great job! I appreciate the effort, guys! :hatched_chick:

**Comment by alongoldboim** [comment]
@tstrachota @mbacovsky mazal tov to all of us :) toke some time, but i like the new feature.

**Pull request merged**